tsdb/index/field_uid.go

```diff
@@ -0,0 +1,227 @@
+	"fmt"
+	"sort"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/eleme/lindb/kv"
+	"github.com/eleme/lindb/pkg/buf"
+	"github.com/eleme/lindb/pkg/constants"
+	"github.com/eleme/lindb/pkg/field"
+	"github.com/eleme/lindb/pkg/logger"
+	"github.com/eleme/lindb/pkg/util"
+)
+
+//FieldUID is unique under the metric name.
+type FieldUID struct {
+	metricMap map[uint32]*FieldIndex //key is metricId
+	family    kv.Family
+	rwLock    sync.RWMutex
+}
+
+//FieldIndex represents
+type FieldIndex struct {
+	fieldMap   map[string]uint32 //key is field name
+	sequenceID uint16            // sequence id
+	rwLock     sync.RWMutex
+}
+
+//FieldReader represents parses byte arrays for easy reading
+type FieldReader struct {
+	reader     *buf.ByteBufReader //byte buf reader
+	seq        uint16             //field sequence id
+	fieldCount int                //field count
+	fieldPos   int                //field start position
+}
+
+func newFieldIndex(seq uint16) *FieldIndex {
+	return &FieldIndex{
+		fieldMap:   make(map[string]uint32),
+		sequenceID: seq,
+	}
+}
+
+func newFieldReader(byteArray []byte) *FieldReader {
+	bufReader := buf.NewBufReader(byteArray)
+	sequence := uint16(bufReader.ReadInt())
+	count := int(bufReader.ReadInt())
+
+	return &FieldReader{
+		reader:     bufReader,
+		seq:        sequence,
+		fieldCount: count,
+		fieldPos:   bufReader.GetPosition(),
+	}
+}
+
+//NewFieldUID creation requires kvFamily
+func NewFieldUID(f kv.Family) *FieldUID {
+	return &FieldUID{
+		metricMap: make(map[uint32]*FieldIndex),
+		family:    f,
+	}
+}
+
+//Finds the ID associated with a given field name or creates it.
+func (f *FieldUID) GetOrCreateFieldID(metricID uint32, fieldName string, fieldType field.Type,
+	create bool) (uint32, error) {
+	fieldIndex := f.metricMap[metricID]
+	if nil == fieldIndex {
+		f.rwLock.Lock()
+		//double check
+		fieldIndex = f.metricMap[metricID]
+		if nil == fieldIndex {
+			fieldIndex = newFieldIndex(f.getLastFieldSequenceID(metricID))
+			f.metricMap[metricID] = fieldIndex
+		}
+		f.rwLock.Unlock()
+	}
+	//if not exists create
+	id := f.GetFieldID(metricID, fieldName)
+	if create {
+		if id == constants.NotFoundFieldID {
+			fieldIndex.sequenceID++
+			fieldIndex.fieldMap[fieldName] = util.ShortToInt(uint16(fieldType), fieldIndex.sequenceID)
+			return fieldIndex.fieldMap[fieldName], nil
+		}
+		return id, nil
+	}
+	return id, nil
+}
+
+//Get all field names within the metric name
+func (f *FieldUID) GetFields(metricID uint32, limit int16) map[string]struct{} {
+	f.getData(metricID, func(fr *FieldReader) bool {
+		for i := 0; i < fr.fieldCount; i++ {
+			_, key := fr.reader.ReadKey()
+			id := fr.reader.ReadInt()
+			//todo
+			fmt.Println("fieldName", key, " id:", id)
+		}
+		return true
+	})
+	return nil
+}
+
+//Forces a flush of in-memory data, and clear it
+func (f *FieldUID) Flush() error {
+	//sort metricIds
+	metricIds := f.getSortMetricIds()
+	//flush fieldId to kv-store
+	flusher := f.family.NewFlusher()
+	f.rwLock.Lock()
```

Comment:
Lock granularity is too large

Reply:
Remove lock dependencies, multithreading is not safe